metadata/src/main/java/org/apache/kafka/controller/MetaLogManager.java

```diff
@@ -0,0 +1,94 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.ApiMessageAndVersion;
+
+/**
+ * The MetaLogManager handles storing metadata and electing leaders.
+ */
+public interface MetaLogManager extends AutoCloseable {
+    /**
+     * Listeners receive notifications from the MetaLogManager.
+     */
+    interface Listener {
+        /**
+         * Called when the MetaLogManager commits a message.
+         *
+         * @param epoch         The controller epoch of the message.
+         * @param index         The index of the message within the controller epoch.
+         * @param message       The message.
+         */
+        void commit(long epoch, long index, ApiMessage message);
+
+        /**
+         * Called when the MetaLogManager has claimed the leadership.
+         *
+         * @param epoch         The controller epoch that is starting.
+         */
+        void claim(long epoch);
+
+        /**
+         * Called when the MetaLogManager has renounced the leadership.
+         *
+         * @param epoch         The controller epoch that has ended.
+         */
+        void renounce(long epoch);
+
+        /**
+         * Called when the MetaLogManager has finished shutting down, and wants to tell its
+         * listener that it is safe to shut down as well.
+         */
+        void beginShutdown();
+    }
+
+    /**
+     * Schedule a write to the log.
+     *
+     * The write will be scheduled to happen at some time in the future.  There is no
+     * error return or exception thrown if the write fails.  Instead, the listener may
+     * regard the write as successful if and only if the MetaLogManager reaches the given
+     * index before renouncing its leadership.
+     *
+     * @param epoch         The controller epoch.
+     * @param message       The message to write and the version to use.
+     *
+     * @return              The index of the message.
+     */
+    long scheduleWrite(long epoch, ApiMessageAndVersion message);
+
+    /**
+     * Renounce the leadership.
+     *
+     * @param epoch         The epoch.  If this does not match the current epoch, this
+     *                      call will be ignored.
+     */
+    void renounce(long epoch);
+
+    /**
+     * Begin shutting down, but don't block.  You must still call close to clean up all
+     * resources.
+     */
+    void beginShutdown();
```

Comment:
Hmm... I'd rather not add more accessor methods just so that we can refer to them in JavaDoc (unless they are useful for some other reason)

I added a sentence to the JavaDoc for close indicating that it is not necessary to call beginShutdown first.  Hopefully that clears up any confusion here.